File: src/p8_rsa.java
import java.util.Scanner;
import java.security.*;
import javax.crypto.Cipher;

public class RSA {
    static void printAsHex(byte[] byteArray) {
        for (byte b: byteArray)
                System.out.print(String.format("%x", b));
            System.out.println();
    }

    public static void main(String[] args) {
        try {
            Scanner scanner = new Scanner(System.in);

            KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance("RSA");
            keyPairGen.initialize(512);
            KeyPair keyPair = keyPairGen.genKeyPair();
            PrivateKey privateKey = keyPair.getPrivate();
            PublicKey publicKey = keyPair.getPublic();

            System.out.print("Enter a message: ");
            String message = scanner.nextLine();
            byte[] messageBytes = message.getBytes("UTF8");

            System.out.println("\nPublic key is");
            printAsHex(publicKey.getEncoded());

            Cipher cipher = Cipher.getInstance("RSA");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);
            byte[] ciphertextBytes = cipher.doFinal(messageBytes);
            System.out.println("\nAfter encryption");
            System.out.println("Ciphertext is");
            printAsHex(ciphertextBytes);

            System.out.println("\n\nPrivate key is");
            printAsHex(privateKey.getEncoded());

            cipher.init(Cipher.DECRYPT_MODE, privateKey);
            byte[] plaintextBytes = cipher.doFinal(ciphertextBytes);
            String plaintext = new String(plaintextBytes);
            System.out.println("\nAfter decryption");
            System.out.println("Plaintext is " + plaintext);

        } catch (Exception e) {
            System.out.println("Error occured");
        }
    }
}

File: src/p8_rsa.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int gcd(int a, int b) {
	int temp;
    while (b != 0) {
        temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int modular_exponentiation(int b, int e, int m) {
    int result = 1;
	b = b % m;
    while (e > 0) {
        if (e % 2 == 1)
            result = (result * b) % m;
        b = (b * b) % m;
        e /= 2;
    }
    return result;
}

int main() {
    int p, q, n, phi_n, e, i, d, m, ci, pi;

    printf("Enter prime number p: ");
    scanf("%d", &p);
    printf("Enter prime number q: ");
    scanf("%d", &q);

    n = p * q;
    phi_n = (p - 1) * (q - 1);

    srand(time(NULL));
    while (1) {
        e = rand() % (phi_n - 1) + 2;
        if (gcd(e, phi_n) == 1)
			break;
    }
    
	for (i = 1; i <= phi_n; i++) {
        if ((e * i) % phi_n == 1)
			d = i;
    }

    printf("n = %d\n", n);
    printf("e = %d\n", e);
    printf("d = %d\n", d);

    printf("Enter message (< %d): ", n);
    scanf("%d", &m);

    printf("After encryption\n");
    ci = modular_exponentiation(m, e, n);
    printf("Ciphertext is %d\n", ci);

    printf("After decryption\n");
    pi = modular_exponentiation(ci, d, n);
    printf("Plaintext is %d\n", pi);

    return 0;
}
